File: 2021/12/03/modify_string.py
"""
https://leetcode.com/problems/replace-all-s-to-avoid-consecutive-repeating-characters/
Author: Andrew Jarombek
Date: 12/3/2021
"""


def modify_string(s: str) -> str:
    chars = list(s)

    for i, c in enumerate(chars):
        if chars[i] == '?':
            if i > 0 and chars[i - 1] == 'a':
                if i + 1 < len(chars) and chars[i + 1] == 'b':
                    chars[i] = 'c'
                else:
                    chars[i] = 'b'
            elif i > 0 and chars[i - 1] == 'b':
                if i + 1 < len(chars) and chars[i + 1] == 'a':
                    chars[i] = 'c'
                else:
                    chars[i] = 'a'
            else:
                if i + 1 < len(chars) and chars[i + 1] == 'a':
                    chars[i] = 'b'
                else:
                    chars[i] = 'a'

    return "".join(chars)

File: 2022/08/30/java/ModifyString.java
/*
 * https://leetcode.com/problems/replace-all-s-to-avoid-consecutive-repeating-characters/
 * @author Andrew Jarombek
 * @since 8/30/2022
 */

public class ModifyString {
    public static String modifyString(String s) {
        char[] chars = s.toCharArray();

        for (int i = 0; i < chars.length; i++) {
            char c = chars[i];
            char replacement = 'a';

            if (c == '?') {
                boolean prev = i == 0 || chars[i - 1] != replacement;
                boolean next = i == chars.length - 1 || chars[i + 1] != replacement;

                while (!prev || !next) {
                    replacement += 1;
                    prev = i == 0 || chars[i - 1] != replacement;
                    next = i == chars.length - 1 || chars[i + 1] != replacement;
                }

                chars[i] = replacement;
            }
        }

        return String.valueOf(chars);
    }

    public static void main(String[] args) {
        assert modifyString("?zs").equals("azs");
        assert modifyString("ubv?w").equals("ubvaw");
        assert modifyString("ubva???d").equals("ubvababd");
    }
}
